### PR TITLE
ISPN-10013 Observability: Distributed tracing

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-10013

Please do not close the issue, this PR is only a partial fix.

Only track Hot Rod cache write requests (i.e. no counters, multimap etc.)
to simplify the implementation.
The implementation is selected with system properties
infinispan.opentracing.factory.class and
infinispan.opentracing.factory.method.
An OpenTracing implementation is not included, instead it must be added
to the server/lib directory.

### CODE REVIEW

File: server/core/src/main/java/org/infinispan/server/core/RequestTracer.java
Comment (ryanemerson) [inline server/core/src/main/java/org/infinispan/server/core/RequestTracer.java:43]: `start()`?

File: server/core/src/main/java/org/infinispan/server/core/RequestTracer.java
Comment (ryanemerson) [inline server/core/src/main/java/org/infinispan/server/core/RequestTracer.java:52]: Why aren't we returning `io.opentracing.Span` here?

We're using the opentracing api to provide implementation independence, so I don't think it's an issue to expose this interface outside of this class. It would make the consumer classes more explicit if they didn't have to pass around instances of `Object`.

File: server/tests/src/test/java/org/infinispan/server/functional/RequestTracingIT.java
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/functional/RequestTracingIT.java:67]: Remove comments? Or is this needed by QE?

File: server/core/src/main/java/org/infinispan/server/core/RequestTracer.java
Comment (danberindei) [reply server/core/src/main/java/org/infinispan/server/core/RequestTracer.java:52]: The return value is `Object` because I developed this first as a patch for the WildFly-based server (https://github.com/infinispan/jdg/pull/1647).

Because API versions 0.31.0 and 0.33.0 are incompatible, putting the API on the boot classpath with the implementation allows us to support both the AppDynamics tracer (which only works with 0.31.0), and the latest Jaeger client (which requires 0.33.0) with the same server. But the server must be able to run without any opentracing-api jar, so I had to make the return type `Object`.

I was hoping in master I could put the implementation jars in `server/lib` instead of modifying the boot classpath, but they're not quite equivalent: `RequestTracer` is in the application classloader, which doesn't have access to the `server/lib` classes.

In this initial PR I made a compromise: I put opentracing-api version 0.31.0 in the server's `/lib` directory. This means an impl that uses 0.31.0 can be just dropped in `/server/lib`, which made the integration test easier to write, but running with the latest Jaeger client requires overwriting the opentracing-api jar in `/lib`.

For the future, I'm considering several options:
* Invoke all opentracing methods via reflection, using the extensions classloader to load interfaces. This would allow everything to stay in `/server/lib`, including opentracing-api.
* Do not include the opentracing-api in `/lib`, but require all the implementation jars to be copied to `/lib`, not `/server/lib`, and extend the server test drivers to make it possible.
* Include a default impl (the latest Jaeger client) in `/lib/jaeger`, but allow users to unpack another impl in `/lib/othertracingimpl` and select it with `-Dopentracing.impl.directory=othertracingimpl`.

File: server/tests/src/test/java/org/infinispan/server/functional/RequestTracingIT.java
Comment (danberindei) [reply server/tests/src/test/java/org/infinispan/server/functional/RequestTracingIT.java:67]: I wanted to make it easy to test with both opentracing 0.31.0 and 0.33.0, although changing the version in the parent POM is still needed for now. Ideally I want the test to run with both impls in the same build.

File: server/core/src/main/java/org/infinispan/server/core/RequestTracer.java
Comment (ryanemerson) [reply server/core/src/main/java/org/infinispan/server/core/RequestTracer.java:52]: Thanks for the detailed explanation :+1:

**Review by ryanemerson — APPROVED** [review]
LGTM assuming CI is good.

@danberindei I couldn't see any docs in the PR. I don't think this should block this PR with it being so close to release, but we should add these in 12.0.1 or 12.1.

**Pull request merged**
